src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+            redfish.configure_fabric_interface(uri, address, static_ip_netmask_str(address), mtusize=static_mtu())
+        time.sleep(20)
+
+    @private
+    def unwire_shelf(self, mgmt_ip):
+        redfish = RedfishClient.cache_get(self.middleware, mgmt_ip)
+        for uri in redfish.fabric_ethernet_interfaces():
+            redfish.configure_fabric_interface(uri, '0.0.0.0', '255.255.255.0', True, mtusize=1500)
+
+    @private
+    async def hardwire_host(self, mgmt_ip, shelf_index, schema, verrors):
+        """Discover which direct links exist to the specified expansion shelf."""
+        # See how many interfaces are available on the expansion shelf
+        shelf_interfaces = await self.middleware.call('jbof.fabric_interface_choices', mgmt_ip)
+        shelf_ip_to_mac = await self.middleware.call('jbof.fabric_interface_macs', mgmt_ip)
+        shelf_ips = list(shelf_ip_to_mac.keys())
+
+        # Setup a dict with the expected IP pairs
+        shelf_ip_to_host_ip = {}
+        for eth_index in range(0, len(shelf_interfaces)):
+            shelf_ip_to_host_ip[jbof_static_ip(shelf_index, eth_index)] = initiator_static_ip(shelf_index, eth_index)
+
+        # Let's check that we have the expected hardwired IPs on the shelf
+        if set(shelf_ips) != set(list(shelf_ip_to_host_ip.keys())):
+            # This should not happen
+            verrors.add(schema, 'JBOF does not have expected IPs.')
+            return
+
+        if await self.middleware.call('failover.licensed'):
+            # HA system
+
+            if not await self.middleware.call('failover.remote_connected'):
+                verrors.add(schema, 'HA system must be in good state to allow configuration.')
+                return
+
+            this_node = await self.middleware.call('failover.node')
+            if this_node not in ['A', 'B']:
+                verrors.add(schema, 'HA system must be in good state to allow configuration: Invalid node')
+                return
+
+            for node in ['A', 'B']:
+                connected_shelf_ips = await self.hardwire_node(node, shelf_index, shelf_ip_to_mac)
+                if not connected_shelf_ips:
+                    # Failed to connect any IPs => error
+                    verrors.add(schema, f'Must be able to communicate with at least one interface on the expansion shelf (node {node}).')
+                    return
+        else:
+            connected_shelf_ips = await self.hardwire_node('', shelf_index, shelf_ip_to_mac)
+            if not connected_shelf_ips:
+                # Failed to connect any IPs => error
+                verrors.add(schema, 'Must be able to communicate with at least one interface on the expansion shelf.')
+                return
+
+    @private
+    async def hardwire_node(self, node, shelf_index, shelf_ip_to_mac):
+        localnode = not node or node == await self.middleware.call('failover.node')
+        shelf_ips = list(shelf_ip_to_mac.keys())
```

Comment:
There is no reason to instantiate this variable at the top of the method when it's not used until you get to line 376. Furthermore, there is no reason to do `list(dict.keys())`. You can remove this line and then on line 375 do
```
for shelf_ip in shelf_ips:
    ...
```